xtask/src/crosstool.rs

```diff
@@ -0,0 +1,227 @@
+use std::fmt::Write;
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use crate::util::{project_dir, write_to_string};
+use clap::Args;
+use cross::ToUtf8;
+
+const DEFAULT_GCC_VERSION: &str = "8.3.0";
+const DEFAULT_GLIBC_VERSION: &str = "2.17.0";
+const DEFAULT_LINUX_VERSION: &str = "4.19.21";
+const DOCKER: &str = "docker";
+const CT_NG: &str = "crosstool-ng";
+const TOOLCHAINS: &str = "cross-toolchains";
+const CT_CONFIG: &str = "crosstool-config";
+
+#[derive(Args, Debug)]
+pub struct ConfigureCrosstool {
+    /// Provide verbose diagnostic output.
+    #[clap(short, long)]
+    pub verbose: bool,
+    /// The gcc version to configure for.
+    #[clap(long, env = "GCC_VERSION")]
+    pub gcc_version: Option<String>,
+    /// The glibc version to configure for.
+    #[clap(long, env = "GLIBC_VERSION")]
+    pub glibc_version: Option<String>,
+    /// The linux version to configure for.
+    #[clap(long, env = "LINUX_VERSION")]
+    pub linux_version: Option<String>,
+    /// Targets to build for
+    #[clap()]
+    targets: Vec<String>,
+}
+
+fn locate_ctng_config(
+    target: &str,
+    root: &Path,
+    cross_toolchains_root: &Path,
+) -> cross::Result<(PathBuf, PathBuf)> {
+    let config_name = format!("{target}.config");
+    let template_name = format!("{config_name}.in");
+    let ct_ng_root = root.join(CT_NG);
+    let cross_toolchains_ctng_root = cross_toolchains_root.join(CT_NG);
+    let (src_root, dst_root) = if cross_toolchains_ctng_root.join(&template_name).exists() {
+        (
+            &cross_toolchains_ctng_root,
+            cross_toolchains_root.join(DOCKER).join(CT_CONFIG),
+        )
+    } else if ct_ng_root.join(&template_name).exists() {
+        (&ct_ng_root, root.join(DOCKER).join(CT_CONFIG))
+    } else {
+        eyre::bail!("unable to find config for target \"{target}\"");
+    };
+    Ok((src_root.join(template_name), dst_root.join(config_name)))
+}
+
+fn read_config_dir(dir: &Path) -> cross::Result<Vec<String>> {
+    let mut targets = vec![];
+    for entry in fs::read_dir(dir)? {
+        let file = entry?;
+        let basename = file.file_name().to_utf8()?.to_string();
+        if let Some(target) = basename.strip_suffix(".config.in") {
+            targets.push(target.to_string());
+        }
+    }
+
+    Ok(targets)
+}
+
+fn configure_target(
```

Comment:
I feel like all the if statements here on major and minor can be combined into a `match (minor, major)` with either range patterns or if guards.